man/man1/printclm.1
.TH printclm 1  "USER COMMANDS"
.SH NAME
printclm, printheader \- Print contents of clm files for LPJmL
.SH SYNOPSIS
.B printclm
[-h] [-metafile] [-header] [-data] [-text] [-json] [-scale] [-longheader] [-type {byte|short|int|float|double}] [\-map name] [\-nbands n] [\-start s] [\-end e] [\-first f] [\-last l] \fIclmfile\fP ...

.B printheader
\fIclmfile\fP ...
.SH DESCRIPTION
Program prints file header and data stored in clm data files for LPJmL. \fBprintheader\fP is an alias for \fBprintclm\fP printing the file header only.
.SH OPTIONS
.TP
\-h,\--help
Print a short help text.
.TP
-metafile
File is a JSON metafile describing the structure of the raw binary file.
.TP
-header
Print without header.
.TP
-data
Print without data.
.TP
-text
Print data without header and cell index.
.TP
-json
Print header in JSON format only.
.TP
-scale
Scale data with scale factor defined in header.
.TP
-longheader
Set long header in CLM files.
.TP
-type {byte|short|int|float|double}
Set datatype of clm file. Default is short.
.TP
-nbands n
Set number of bands to n ignoring value in file header.
.TP
-map name
Name of map in JSON metafile, default is "map".
.TP
-start s
Start output of data at year s.
.TP
-end e
End output of data at year e.
.TP
-first f
Index of first cell for output.
.TP
-last l
Index of last cell for output.
.TP
.I clmfile
Filename of clm data file(s).
.SH EXAMPLE
.TP
Print headers of all clm files in the current directory:
.B printclm
-data *.clm
.TP
or
.B printheader
*.clm
.PP
.SH EXIT STATUS
.B printclm
and
.B printheader
returns a zero exit status if clm file could be printed.
Non zero is returned in case of failure.

.SH AUTHORS

For authors and contributors see AUTHORS file

.SH COPYRIGHT

(C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file

.SH SEE ALSO
setclm(1), statclm(1), headersize(1), lpjml(1), cru2clm(1), grid2clm(1), cft2clm(1), clm2cdf(1), cdf2clm(1), clm(5)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project follows a [GNU standard version numbering](https://www.gnu.org/prep/standards/html_node/Releases.html#index-version-numbers_002c-for-releases)
of `major.minor.patch` with
- `major` for substantial changes to code functionality,
- `minor` for additions of modules and functionality, and
- `patch` for technical changes and bugfixes without extending the functionality of the code.

- if applicable, categories to include in the CHANGELOG.md are:
  - Added
  - Changed
  - Deprecated
  - Removed
  - Fixed
  - Security

## [Unreleased]


### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de)

### Added

- Option `-h` added to utilities `addheader`, `cdf2bin`, `regridclm`, `regridsoil` and `printclm` to print help text.
- Notice added in `configure.sh` that `make clean; make all` has to be performed after change in configuration.

### Changed

- If option `-ncell 0` is used in `addheader` utility then the number of cells is calculated from the file size of the binary file.
- Utilities `grid2clm` and `cft2clm` are replaced by alias to `addheader`.

### Fixed

- Missing `free(axis->comment)` added in `freeaxis()` to avoid memory leak.
- Missing deallocation of list added in `open_config.c`.
- Function `nc_close()` replaced by `closeclimate_netcdf()` in `cdf2clm.c` and `cdf2bin.c` to avoid memory leak.
- Argument for error message corrected in `getcellindex.c` and `getcountry.c`.
- Typo in man page of `configure.sh` fixed.


## [5.9.25] - 2025-05-22

### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de)
- code review: Stephen Wirth (wirth@pik-potsdam.de), Fabian Stenzel (stenzel@pik-potsdam.de), Sebastian Ostberg(ostberg@pik-potsdam.de)

### Changed

- Axis names and missing values for NetCDF output are read from JSON file `netcdf.cjson` instead of being hard coded.
- Warnings in `addheader` replaced by terminating errors.
- Function `openinputfile()` returns an error if cell size differs in JSON metafile.
- Function `openmetafile()` returns an error if binary file is empty.
- Output format for cell size updated in `fprintheader.c`, `fprintoutputjson.c`, `fprintjson.c`, and `openinputfile.c`.

### Added

- New utility `splitclm` added to copy specific bands from a clm file.
- Environment variables `LPJNOPP` and `LPJPEDANTIC` added. If set to `true`, preprocessor is disabled and pedantic checking is enabled, respectively.
- Info about `--help` option added in case of error in command line arguments.
- Option `-config` added to `cdf2clm`, `bin2cdf`, `country2cdf` and `clm2cdf` in order to read axis names and missing values from a JSON file. File `netcdf.json` added as a template.
- Option `-json` added to `arr2clm`, `asc2clm`, `txt2grid`, `txt2clm`, `getcountry`, and `cdf2coord` in order to create additional JSON metafiles.
- Option `-inpath` added to `regridlpj` in order to set directory where data can be found.
- Options `-zero` and `-search` added to `regridlpj`.
- Option `-double` added to `txt2grid` in order to create grid files with coordinate values of type double.
- Option `-timestep` added to `addheader`, `txt2clm` and `bin2cdf`.
- Option `sum`, `tsum` and `tmean` to `mathclm` added.
- Check for correct setting of nstep and nbands added in `txt2clm`.

### Fixed

- Misspelled `-fsanitize=address` option corrected in `Makefile.mpich`.
- Missing `free(cmd)` added in `fprintincludes.c`.
- Parsing of options corrected in `fprintincludes.c`.
- Utility `bin2cdf` fixed for timestep>1.


## [5.9.24] - 2025-05-15

### Contributors

- author: David Hötten (davidho@pik-potsdam.de)
- code review: Sebastian Ostberg (ostberg@pik-potsdam.de), Maik Billing (billing@pik-potsdam.de)

### Added

- Merge request template ``.gitlab/merge_request_templates/default.md`` that includes several checklists.


## [5.9.23] - 2025-05-08

### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de)
- code review: Jens Heinke (heinke@pik-potsdam.de), Fabian Stenzel(stenzel@pik-potsdam.de)

### Changed

- Option `-check` of `configure.sh` enables now run-time checks of memory leaks and memory access out of bounds and undefined variables for gcc and icx compiler. Optimization is not disabled. Use options `-debug -check` to disable.
- Datatype `List` used in `fprintfiles.c`.
- All allocated memory is now freed in utilities.
- Prescribed landcover input file changed in `input.cjson` to a file with the same number of natural PFTs.

src/utils/printclm.c
/**************************************************************************************/
/**                                                                                \n**/
/**                          p  r  i  n  t  c  l  m  .  c                          \n**/
/**                                                                                \n**/
/**     Print contents of *.clm files                                              \n**/
/**                                                                                \n**/
/** (C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file    \n**/
/** authors, and contributors see AUTHORS file                                     \n**/
/** This file is part of LPJmL and licensed under GNU AGPL Version 3               \n**/
/** or later. See LICENSE file or go to http://www.gnu.org/licenses/               \n**/
/** Contact: https://github.com/PIK-LPJmL/LPJmL                                    \n**/
/**                                                                                \n**/
/**************************************************************************************/

#include "lpj.h"

#define USAGE "Usage: %s [-h] [-v] [-metafile] [-header] [-data] [-text] [-json] [-scale]\n       [-longheader] [-type {byte|short|int|float|double}] [-map name]\n       [-nbands n] [-start s] [-end e] [-first f] [-last l] filename ...\n"
#define ERR_USAGE USAGE "\nTry \"%s --help\" for more information.\n"

#define NO_HEADER 1
#define NO_DATA 2
#define NO_TEXT 4

static void printclm(const char *filename,int output,int nbands,int version,
                     int start,int stop,int first,int last,Type type,Bool ismeta,const char *map_name,Bool isscale,Bool isjon)
{
  FILE *file;
  time_t mod_date;
  long long size;
  Header header;
  Restartheader restartheader;
  String id;
  short sdata;
  int idata;
  char byte;
  float fdata;
  double ddata;
  int year,cell,i,*index=NULL,rc,t;
  char *unit=NULL,*long_name=NULL,*standard_name=NULL;
  Bool swap,isrestart,isreservoir;
  size_t offset;
  Reservoir reservoir;
  Map *map=NULL;
  Attr *attrs=NULL;
  int n_attr=0;
  if(ismeta)
  {
    isrestart=isreservoir=FALSE;
    header.scalar=1;
    header.cellsize_lon=header.cellsize_lat=0.5;
    header.firstyear=1901;
    header.firstcell=0;
    header.nyear=1;
    header.nbands=(nbands==-1)  ? 1 : nbands;
    header.nstep=1;
    header.timestep=1;
    header.datatype=type;
    header.order=CELLYEAR;
    file=openmetafile(&header,&map,map_name,&attrs,&n_attr,NULL,NULL,NULL,&unit,&standard_name,&long_name,NULL,NULL,NULL,&swap,&offset,filename,TRUE);
    if(file==NULL)
      return;
    if(fseek(file,offset,SEEK_CUR))
    {
      fprintf(stderr,"Error seeking in '%s' to offset %lu.\n",filename,offset);
      fclose(file);
      return;
    }
    type=header.datatype;
    version=CLM_MAX_VERSION+1;
    strcpy(id,"description file");
  }
  else
  {
    unit=NULL;
    standard_name=NULL;
    long_name=NULL;
    file=fopen(filename,"rb");
    if(file==NULL)
    {
      fprintf(stderr,"Error opening '%s': %s.\n",filename,strerror(errno));
      return;
    }
    if(freadanyheader(file,&header,&swap,id,&version,TRUE))
    {
      fprintf(stderr,"Error reading header in '%s', perhaps not a clm file.\n",
              filename);
      return;
    }
    if(version>2)
      type=header.datatype;
    else
      header.datatype=type;
    if(nbands!=-1)
      header.nbands=nbands;
    isrestart=(!strcmp(id,RESTART_HEADER));
    isreservoir=(!strcmp(id,LPJRESERVOIR_HEADER));
  }
  if(isjon)
  {
    fprintjson(stdout,filename,NULL,NULL,NULL,NULL,&header,map,map_name,attrs,n_attr,NULL,unit,standard_name,long_name,NULL,LPJ_SHORT,CLM,id,swap,version);
    return;
  }
  free(long_name);
  freemap(map);
  freeattrs(attrs,n_attr);
  if((output & NO_HEADER)==0)
  {
    mod_date=getfiledate(filename);
    printf("Created:\t%s"
           "Id:\t\t%s\n"
           "Version:\t%d\n"
           "Byte order:\t",
           asctime(localtime(&mod_date)),id,version);
    if(bigendian())
      printf((swap) ? "Little endian" : "Big endian");
    else
      printf((swap) ? "Big endian" : "Little endian");
    putchar('\n');
    printheader(&header);
    if(map!=NULL)
    {
      printf("%s: ",map_name);
      printmap(map);
      printf("\n");
      freemap(map);
    }
    if(unit!=NULL && strlen(unit)>0)
      printf("Unit:\t\t%s\n",unit);
    if(isrestart)
    {
      if(RESTART_VERSION==version)
      {
        freadrestartheader(file,&restartheader,swap);
        printf("Land use:\t\t%s\n"
               "River routing:\t\t%s\n"
               "Fixed sowing date:\t%s\n"
               "Prescribed PHU:\t\t%s\n"
               "Double harvest:\t\t%s\n",
               bool2str(restartheader.landuse),
               bool2str(restartheader.river_routing),
               bool2str(restartheader.sdate_option),
               bool2str(restartheader.crop_phu_option),
               bool2str(restartheader.separate_harvests));
        printf("Random seed:\t");
        for(i=0;i<NSEED;i++)
          printf(" %d",restartheader.seed[i]);
        putchar('\n');
      }
      else
        fprintf(stderr,"Warning: invalid restart version %d, must be %d.\n",
                version,RESTART_VERSION);
    }
  }
  free(unit);
  if(!ismeta && !isrestart && version>CLM_MAX_VERSION)
    fprintf(stderr,"Warning: Unsupported version %d, must be less than %d.\n",
            version,CLM_MAX_VERSION+1);
  if(!ismeta && header.order==CELLINDEX)
  {
    if(version<3)
      type=LPJ_FLOAT;
    size=getfilesizep(file)-headersize(id,version)-sizeof(int)*header.ncell;
    if(size!=(long long)typesizes[type]*header.ncell*header.nbands*header.nstep*header.nyear)
      fprintf(stderr,"Warning: File length does not match header, differs by %lld bytes.\n",
              llabs(size-(long long)typesizes[type]*header.ncell*header.nbands*header.nstep*header.nyear));
  }
  else if(!isrestart && !isreservoir && !ismeta)
  {
    size=getfilesizep(file)-headersize(id,version);
    if(size!=typesizes[type]*header.ncell*header.nbands*header.nyear*header.nstep)
      fprintf(stderr,"Warning: File length does not match header, differs by %lld bytes.\n",
              llabs(size-(long long)typesizes[type]*header.ncell*header.nbands*header.nstep*header.nyear));
  }
  if((output & NO_DATA)==0)
  {
    if(last==INT_MAX)
      last=header.ncell-first;
    else
      last-=first-1;
    if(isrestart)
      fputs("File is restart file, use 'lpjprint' to print content.\n",stderr);
    else if(isreservoir)
    {
      if((output & NO_TEXT)==0)
        puts("Cell  Year Capacity    Area   Inst. cap. Height Purpose\n"
             "----- ---- ----------- ------ ---------- ------ ---------");
      if(fseek(file,sizeof(Reservoir)*(first-header.firstcell),SEEK_CUR))
      {
        fprintf(stderr,"Error seeking in '%s' to position %d.\n",filename,first);
        fclose(file);
        return;
      }
      for(cell=0;cell<last;cell++)
      {
        if(readreservoir(&reservoir,swap,file))
        {
          fprintf(stderr,"Unexpected end of file in '%s' at position %d.\n",
                  filename,cell);
          break;
        }
        if(reservoir.year>0)
        {
          printf("%5d %4d %11g %6g %10d %6d",
                 cell,reservoir.year,reservoir.capacity,reservoir.area,
                 reservoir.inst_cap,reservoir.height);
          for(i=0;i<NPURPOSE;i++)
            printf(" %1d",reservoir.purpose[i]);
          putchar('\n');
        }
      } /* of for(cell=...) */
    }
    else
    {
      if(header.nyear<=0)
      {
        fprintf(stderr,"Invalid nyear=%d, set to one.\n",header.nyear);
        header.nyear=1;
      }
      if(start==INT_MAX)
        start=header.firstyear;
      else if(start<header.firstyear)
      {
        fprintf(stderr,"Invalid first year %d, set to %d.\n",start,header.firstyear);
        start=header.firstyear;
      }
      else if(start>header.firstyear+header.nyear-1)
      {
        fprintf(stderr,"Invalid first year %d>%d, no output written.\n",start,header.firstyear+header.nyear-1);
        fclose(file);
        return;
      }
      if(stop==INT_MAX)
        stop=header.firstyear+header.nyear-1;
      else if(stop>header.firstyear+header.nyear-1)
      {
        fprintf(stderr,"Invalid end year %d set to %d.\n",
                start,header.firstyear+header.nyear-1);
        stop=header.firstyear+header.nyear-1;
      }
      if(header.order==CELLINDEX)
      {
        fseek(file,sizeof(int)*(first-header.firstcell),SEEK_CUR);
        index=newvec(int,last);
        if(index==NULL)
        {
          printallocerr("index");
          return;
        }
        rc=freadint(index,last,swap,file);
        if(rc!=last)
        {
          fprintf(stderr,"Unexpected end of file at cell %d.\n",rc+first+1);
          free(index);
          return;
        }
        fseek(file,sizeof(int)*(header.ncell-last-first+header.firstcell),SEEK_CUR);
      }
      if(fseek(file,typesizes[type]*header.ncell*header.nstep*header.nbands*(start-header.firstyear),SEEK_CUR))
      {
        fprintf(stderr,"Error seeking to year %d.\n",start);
        fclose(file);
        return;
      }
      if(header.order==CELLINDEX || header.order==CELLSEQ)
      {
        for(year=start;year<=stop;year++)
          for(t=0;t<header.nstep;t++)
          for(i=0;i<header.nbands;i++)
          {
            if((output & NO_TEXT)==0)
              printf("Year, Step, Band: %d %d %d\n",year,t,i);
            fseek(file,typesizes[type]*(first-header.firstcell),SEEK_CUR);
            for(cell=0;cell<last;cell++)
            {
              if((output & NO_TEXT)==0)
              {
                if(header.order==CELLINDEX)
                  printf("%5d(%5d): ",cell+first,index[cell]);
                else
                  printf("%5d: ",cell);
              }
              switch(type)
              {
                case LPJ_BYTE:
                  if(fread(&byte,1,1,file)!=1)
                  {
                    fprintf(stderr,"Unexpected end of file at year %d, cell %d, band %d.\n",
                            year,cell,i);
                    return;
                  }
                  if(isscale && header.scalar!=1)
                    printf("%6g\n",byte*header.scalar);
                  else
                    printf("%3d\n",(int)byte);
                  break;
                case LPJ_SHORT:
                  if(freadshort(&sdata,1,swap,file)!=1)
                  {
                    fprintf(stderr,"Unexpected end of file at year %d, cell %d, band %d.\n",
                            year,cell,i);
                    return;
                  }
                  if(isscale && header.scalar!=1)
                    printf("%6g\n",sdata*header.scalar);
                  else
                    printf("%5d\n",(int)sdata);
                  break;
                case LPJ_INT:
                  if(freadint(&idata,1,swap,file)!=1)
                  {
                    fprintf(stderr,"Unexpected end of file at year %d, cell %d, band %d.\n",
                            year,cell,i);
                    return;
                  }
                  if(isscale && header.scalar!=1)
                    printf("%6g\n",idata*header.scalar);
                  else
                    printf("%6d\n",idata);
                  break;
                case LPJ_FLOAT:
                  if(freadfloat(&fdata,1,swap,file)!=1)
                  {
                    fprintf(stderr,"Unexpected end of file at year %d, cell %d, day %d.\n",
                            year,cell,i);
                    return;
                  }
                  printf("%g\n",fdata);
                  break;
                case LPJ_DOUBLE:
                  if(freaddouble(&ddata,1,swap,file)!=1)
                  {
                    fprintf(stderr,"Unexpected end of file at year %d, cell %d, band %d.\n",
                            year,cell,i);
                    return;
                  }
                  printf("%6g\n",ddata);
                  break;
               }
            }
            fseek(file,typesizes[type]*(header.ncell-last-first+header.firstcell),SEEK_CUR);
          }
        if(header.order==CELLINDEX)
          free(index);
      }
      else
        for(year=start;year<=stop;year++)
        {
        if((output & NO_TEXT)==0)
          printf("Year: %d\n",year);
        fseek(file,typesizes[type]*header.nbands*header.nstep*(first-header.firstcell),SEEK_CUR);
        for(cell=0;cell<last;cell++)
        {
          if((output & NO_TEXT)==0)
            printf("%5d:",cell+first);
          for(i=0;i<header.nbands*header.nstep;i++)
            switch(type)
            {
              case LPJ_BYTE:
                if(fread(&byte,1,1,file)!=1)
                {
                  putchar('\n');
                  fprintf(stderr,"Unexpected end of file at year %d, cell %d, band %d.\n",
                          year,cell,i);
                  return;
                }
                if(isscale && header.scalar!=1)
                  printf(" %6g",byte*header.scalar);
                else
                  printf(" %3d",(int)byte);
                break;
              case LPJ_SHORT:
                if(freadshort(&sdata,1,swap,file)!=1)
                {
                  putchar('\n');
                  fprintf(stderr,"Unexpected end of file at year %d, cell %d, band %d.\n",
                          year,cell,i);
                  return;
                }
                if(isscale && header.scalar!=1)
                  printf(" %6g",sdata*header.scalar);
                else
                  printf(" %5d",(int)sdata);
                break;
              case LPJ_INT:
                if(freadint(&idata,1,swap,file)!=1)
                {
                  putchar('\n');
                  fprintf(stderr,"Unexpected end of file at year %d, cell %d, band %d.\n",
                          year,cell,i);
                  return;
                }
                if(isscale && header.scalar!=1)
                  printf(" %6g",idata*header.scalar);
                else
                  printf(" %6d",idata);
                break;
              case LPJ_FLOAT:
                if(freadfloat(&fdata,1,swap,file)!=1)
                {
                  putchar('\n');
                  fprintf(stderr,"Unexpected end of file at year %d, cell %d, band %d.\n",
                          year,cell,i);
                  return;
                }
                printf(" %10.6f",fdata);
                break;
              case LPJ_DOUBLE:
                if(freaddouble(&ddata,1,swap,file)!=1)
                {
                  putchar('\n');
                  fprintf(stderr,"Unexpected end of file at year %d, cell %d, band %d.\n",
                          year,cell,i);
                  return;
                }
                printf(" %6g",ddata);
                break;
            }
          putchar('\n');
        } /* of for(cell=...) */
        fseek(file,typesizes[type]*header.nbands*header.nstep*(header.ncell-last-first+header.firstcell),SEEK_CUR);
      }
    }
  }
  fclose(file);
} /* of 'printclm' */

int main(int argc,char **argv)
{
  int i,output,index,start,stop,first,last,nbands,version,rc;
  Type type;
  const char *progname;
  char *endptr;
  char *map_name;
  Bool ismeta;
  Bool isscale;
  Bool isjson;
  progname=strippath(argv[0]);
  output=0;
  first=0;
  start=stop=last=INT_MAX;
  type=LPJ_SHORT;
  nbands=-1;
  ismeta=isscale=isjson=FALSE;
  version=READ_VERSION;
  map_name=MAP_NAME;
  for(i=1;i<argc;i++)
    if(argv[i][0]=='-')
    {
      if(!strcmp(argv[i],"-h") || !strcmp(argv[i],"--help"))
      {
        printf("     ");
        rc=printf("%s (" __DATE__ ") Help",progname);
        printf("\n     ");
        repeatch('=',rc);
        printf("\n\nPrint contents of clm files for LPJmL version %s\n",getversion());
        printf(USAGE
               "\nArguments:\n"
               "-metafile   file is a JSON metafile describing the structure of the raw binary file\n"
               "-header     print without header\n"
               "-data       print without data\n"
               "-text       print data without header and cell index\n"
               "-json       print header in JSON format only\n"
               "-scale s    scale data with scale factor defined in header\n"
               "-longheader set long header(version 2) in CLM files\n"
               "-type t     set datatype of clm file. Default is short\n"
               "-nbands n   set number of bands to n ignoring value in file header\n"
               "-map name   nme of map in JSON metafile, default is \"map\"\n"
               "-start s    start output of data at year s\n"
               "-end e      end output of data at year e\n"
               "-first f    index of first cell for output\n"
               "-last l     index of last cell for output\n"
               "clmfile     filename of clm data file(s)\n\n"
               "(C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file\n",
               progname);
        return EXIT_SUCCESS;
      }
      else if(!strcmp(argv[i],"-v") || !strcmp(argv[i],"--version"))
      {
        puts(getversion());
        return EXIT_SUCCESS;
      }
      else if(!strcmp(argv[i],"-header"))
        output|=NO_HEADER;
      else if(!strcmp(argv[i],"-data"))
        output|=NO_DATA;
      else if(!strcmp(argv[i],"-text"))
        output|=NO_HEADER|NO_TEXT;
      else if(!strcmp(argv[i],"-json"))
        isjson=TRUE;
      else if(!strcmp(argv[i],"-longheader"))
        version=2;
      else if(!strcmp(argv[i],"-metafile"))
        ismeta=TRUE;
      else if(!strcmp(argv[i],"-scale"))
        isscale=TRUE;
      else if(!strcmp(argv[i],"-map"))
      {
        if(argc-1==i)
        {
          fprintf(stderr,"Argument missing for option '-map'.\n"
                  ERR_USAGE,progname,progname);
          return EXIT_FAILURE;
        }
        map_name=argv[++i];
      }
      else if(!strcmp(argv[i],"-first"))
      {
        if(argc-1==i)
        {
          fprintf(stderr,"Argument missing for option '-first'.\n"
                  ERR_USAGE,progname,progname);
          return EXIT_FAILURE;
        }
        first=strtol(argv[++i],&endptr,10);
        if(*endptr!='\0')
        {
          fprintf(stderr,"Invalid number '%s' for option '-first'.\n"
                  ERR_USAGE,argv[i],progname,progname);
          return EXIT_FAILURE;
        }
      }
      else if(!strcmp(argv[i],"-last"))
      {
        if(argc-1==i)
        {
          fprintf(stderr,"Argument missing for option '-last'.\n"
                  ERR_USAGE,progname,progname);
          return EXIT_FAILURE;
        }
        last=strtol(argv[++i],&endptr,10);
        if(*endptr!='\0')
        {
          fprintf(stderr,"Invalid number '%s' for option '-last'.\n"
                  ERR_USAGE,argv[i],progname,progname);
          return EXIT_FAILURE;
        }
      }
      else if(!strcmp(argv[i],"-start"))
      {
        if(argc-1==i)
        {
          fprintf(stderr,"Argument missing for option '-start'.\n"
                  ERR_USAGE,progname,progname);
          return EXIT_FAILURE;
        }
        start=strtol(argv[++i],&endptr,10);
        if(*endptr!='\0')
        {
          fprintf(stderr,"Invalid number '%s' for option '-start'.\n"
                  ERR_USAGE,argv[i],progname,progname);
          return EXIT_FAILURE;
        }
      }
      else if(!strcmp(argv[i],"-nbands"))
      {
        if(argc-1==i)
        {
          fprintf(stderr,"Argument missing for option '-nbands'.\n"
                  ERR_USAGE,progname,progname);
          return EXIT_FAILURE;
        }
        nbands=strtol(argv[++i],&endptr,10);
        if(*endptr!='\0')
        {
          fprintf(stderr,"Invalid number '%s' for option '-nbands'.\n"
                  ERR_USAGE,argv[i],progname,progname);
          return EXIT_FAILURE;
        }
        if(nbands<1)
        {
          fputs("Invalid value for number of bands, must be >0.\n",stderr);
          return EXIT_FAILURE;
        }
      }
      else if(!strcmp(argv[i],"-end"))
      {
        if(argc-1==i)
        {
          fprintf(stderr,"Argument missing for option '-end'.\n"
                  ERR_USAGE,progname,progname);
          return EXIT_FAILURE;
        }
        stop=strtol(argv[++i],&endptr,10);
        if(*endptr!='\0')
        {
          fprintf(stderr,"Invalid number '%s' for option '-end'.\n"
                  ERR_USAGE,argv[i],progname,progname);
          return EXIT_FAILURE;
        }
        if(start!=INT_MAX && stop<start)
        {
          fprintf(stderr,"Error: argument of '-end' must be  >=%d.\n",start);
          return EXIT_FAILURE;
        }
      }
      else if(!strcmp(argv[i],"-type"))
      {
        if(argc-1==i)
        {
          fprintf(stderr,"Argument missing for option '-type'.\n"
                  ERR_USAGE,progname,progname);
          return EXIT_FAILURE;
        }
        index=findstr(argv[++i],typenames,N_TYPES);
        if(index==NOT_FOUND)
        {
          fprintf(stderr,"Invalid argument '%s' for option '-type'.\n"
                  ERR_USAGE,argv[i],progname,progname);
          return EXIT_FAILURE;
        }
        type=(Type)index;
      }
      else
      {
        fprintf(stderr,"Invalid option '%s'.\n"
                ERR_USAGE,argv[i],progname,progname);
        return EXIT_FAILURE;
      }
    }
    else
      break;
  argc-=i;
  argv+=i;
  if(argc==0)
  {
    fprintf(stderr,"Error: Filename missing.\n"
            ERR_USAGE,progname,progname);
    return EXIT_FAILURE;
  }
  for(i=0;i<argc;i++)
  {
    if(argc>1)
      printf("Filename:\t%s\n",argv[i]);
    printclm(argv[i],output,nbands,version,start,stop,first,last,type,ismeta,map_name,isscale,isjson);
  }
  return EXIT_SUCCESS;
} /* of 'main' */
